File: web/backend/app/websocket_manager.py
from __future__ import annotations

import asyncio
from collections import defaultdict

from fastapi import WebSocket


class WebSocketManager:
    def __init__(self) -> None:
        self._connections: dict[str, set[WebSocket]] = defaultdict(set)
        self._loop: asyncio.AbstractEventLoop | None = None

    async def connect(self, job_id: str, websocket: WebSocket) -> None:
        await websocket.accept()
        self._loop = asyncio.get_running_loop()
        self._connections[job_id].add(websocket)

    def disconnect(self, job_id: str, websocket: WebSocket) -> None:
        if job_id in self._connections:
            self._connections[job_id].discard(websocket)
            if not self._connections[job_id]:
                self._connections.pop(job_id, None)

    def broadcast(self, job_id: str, payload: dict[str, object]) -> None:
        sockets = list(self._connections.get(job_id, set()))
        if not sockets or self._loop is None:
            return
        asyncio.run_coroutine_threadsafe(self._broadcast_async(job_id, payload), self._loop)

    async def _broadcast_async(self, job_id: str, payload: dict[str, object]) -> None:
        for websocket in list(self._connections.get(job_id, set())):
            try:
                await websocket.send_json(payload)
            except Exception:
                self.disconnect(job_id, websocket)

File: kotlin/terrain-web/src/main/kotlin/com/terrainconverter/web/WebSocketManager.kt
package com.terrainconverter.web

import io.ktor.websocket.Frame
import io.ktor.websocket.WebSocketSession
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.util.Collections
import java.util.concurrent.ConcurrentHashMap

class WebSocketManager(private val json: Json) {
    private val connections = ConcurrentHashMap<String, MutableSet<WebSocketSession>>()

    fun connect(jobId: String, session: WebSocketSession) {
        connections.computeIfAbsent(jobId) { Collections.synchronizedSet(linkedSetOf()) }.add(session)
    }

    fun disconnect(jobId: String, session: WebSocketSession) {
        connections[jobId]?.let {
            it.remove(session)
            if (it.isEmpty()) connections.remove(jobId)
        }
    }

    suspend fun broadcast(jobId: String, payload: String) {
        val sockets = connections[jobId]?.toList() ?: return
        sockets.forEach { socket ->
            try {
                socket.send(Frame.Text(payload))
            } catch (_: Exception) {
                disconnect(jobId, socket)
            }
        }
    }

    suspend fun broadcastJob(job: JobDetail) {
        broadcast(job.id, json.encodeToString(JobEvent(job = job)))
    }

    suspend fun broadcastLog(jobId: String, line: String) {
        broadcast(jobId, json.encodeToString(LogEvent(line = line)))
    }
}
